// react-flow-renderer/dist/utils/index.js
export const getDimensions = (node) => ({
    width: node.offsetWidth,
    height: node.offsetHeight,
});
export const clamp = (val, min = 0, max = 1) => Math.min(Math.max(val, min), max);
export const clampPosition = (position, extent) => ({
    x: clamp(position.x, extent[0][0], extent[1][0]),
    y: clamp(position.y, extent[0][1], extent[1][1]),
});
export const getHostForElement = (element) => element.getRootNode?.() || window?.document;
export const getBoundsOfBoxes = (box1, box2) => ({
    x: Math.min(box1.x, box2.x),
    y: Math.min(box1.y, box2.y),
    x2: Math.max(box1.x2, box2.x2),
    y2: Math.max(box1.y2, box2.y2),
});
export const rectToBox = ({ x, y, width, height }) => ({
    x,
    y,
    x2: x + width,
    y2: y + height,
});
export const boxToRect = ({ x, y, x2, y2 }) => ({
    x,
    y,
    width: x2 - x,
    height: y2 - y,
});
export const getBoundsofRects = (rect1, rect2) => boxToRect(getBoundsOfBoxes(rectToBox(rect1), rectToBox(rect2)));
export const isNumeric = (n) => !isNaN(n) && isFinite(n);
